### PR TITLE
differenciate cross-build case to use nostd sources

### LINKED ISSUE #9111: Rust Cross compile to embedded target fails due to non working rustc
**Describe the bug**
Trying to cross compile with the rust language to an embedded target (no std support), fails at the setup step.

I think adding `#![no_std]` to https://github.com/mesonbuild/meson/blob/a9e9b7c7501a3c8a5984a93879d1f309bf8c72aa/mesonbuild/compilers/rust.py#L69-L72 would solve the issue.

**To Reproduce**

meson.build:
```
project('bare-metal', ['rust'])

static_library('baremetal', 'src/lib.rs', rust_crate_type: 'staticlib')
```

cross.ini
```
[binaries]
c = '/usr/bin/arm-none-eabi-gcc'
rust_ld = '/usr/bin/arm-none-eabi-gcc'
rust = ['rustc', '--target', 'thumbv7m-none-eabi']

[host_machine]
system = 'lm3s6965evb'
cpu_family = 'arm'
cpu = 'thumbv7'
endian = 'little'
```

**Expected behavior**
It compiles.

### PR DESCRIPTION
**Pull request by PThierry**
In cross-compilation use cases, some target do not support libstd and require rustc sanity check to test only the bare-metal compilation mode.

Fixes #9111

### CODE REVIEW

**Comment by xclaesse** [comment]
A no-std binary seems indeed very complicated, found that on the topic: https://fasterthanli.me/series/making-our-own-executable-packer/part-12.

But in the case the cross build has an exe_wrapper, it's going to try to run the executable. We probably still should have a main() in the case we have an exe_wrapper.

Maybe  there is a way to detect if we have the std available? That could be needed to deal with `self.native_static_libs`.

**Comment by xclaesse** [comment]
FWIW, found that this works:
```rust
#![no_std]
#![no_main]
#[no_mangle]
pub fn main() {
}

#[panic_handler]
fn panic(_info: &core::panic::PanicInfo) -> ! {
    loop {}
}
```

But you have to add `-C panic=abort -C default-linker-libraries`. Not sure it's worth the trouble...

**Comment by PThierry** [comment]
Okay found the way, after tooking some time reading https://os.phil-opp.com/freestanding-rust-binary/
this requires the nostartfile, avoiding libc linking with a non-existing _exit() extern symbol resolution trying, which is a non-sense in bare metal (although bare netal only, not all  cross use cases :-/)
Differenciating various cross use case would require something different, to detect that the project is a bare metal target (kernel, firmware...)

**Comment by xclaesse** [comment]
@PThierry That's a bit better variant to what I posted above indeed, does not even need libc in that case. But that's an infinite loop, right? In the case you build with an exe_wrapper (e.g. wine, qemu) that would deadlock.

**Comment by PThierry** [comment]
> @PThierry That's a bit better variant to what I posted above indeed, does not even need libc in that case. But that's an infinite loop, right? In the case you build with an exe_wrapper (e.g. wine, qemu) that would deadlock.

Ack ! fixed.

**Comment by xclaesse** [comment]
IIRC, I tried that at some point but it segfault because there is no exit(). The link I posted earlier has an asm implementation for that, but that not portable I imagine. 

I think it would be safer to have a test that check if we have the std. If yes, do a normal main(), otherwise just build a library and skip running it.

**Comment by PThierry** [comment]
I moved to a 'test and fallback'  method, which first try a nominal build (with libstd) even in cross build mode, and if it fails, fallback to a bare metal case. This allows to, if the bare-metal build succeeds, skip the exe_wrapper only in that case and not others (cross with libstd typically).

**Comment by PThierry** [comment]
Hello @xclaesse @dcbaker ! Is the current PR state okay or do you think there is still something missing in the current patch implementation ? :-)
Thanks !

**Comment by tristan957** [comment]
Wanna rebase the PR? Targeting for 1.3.0

**Comment by tristan957** [comment]
Also squash commits where appropriate.

File: mesonbuild/compilers/rust.py [hunk 1]
```diff
@@ -75,22 +75,47 @@ def needs_static_linker(self) -> bool:
     def sanity_check(self, work_dir: str, environment: 'Environment') -> None:
         source_name = os.path.join(work_dir, 'sanity.rs')
         output_name = os.path.join(work_dir, 'rusttest')
+        is_bare = False
+        optional_args = []
+        # initial tesst: nominal build with std library
```
Comment (tristan957) [inline mesonbuild/compilers/rust.py:80]: test

Comment (PThierry) [reply mesonbuild/compilers/rust.py:80]: Ack, rebase and squashing :)

**Comment by PThierry** [comment]
squash & rebase done :-)

**Comment by tristan957** [comment]
@xclaesse anything else you need from this PR?

**Comment by PThierry** [comment]
rebased to last master and fixed conflict

**Comment by fvalette-ledger** [comment]
> A no-std binary seems indeed very complicated, found that on the topic: https://fasterthanli.me/series/making-our-own-executable-packer/part-12.
> 
> But in the case the cross build has an exe_wrapper, it's going to try to run the executable. We probably still should have a main() in the case we have an exe_wrapper.
> 
> Maybe there is a way to detect if we have the std available? That could be needed to deal with `self.native_static_libs`.

May be you could have a look to #13486. IMHO, building no-std means having `self.info.kernel` to `none`.
In this case, `self.native_static_libs` is empty (i.e. `rustc` is silent).
We are using a workaround since this PR is opened, but was broken since `1.5.0` (no-std rust static lib in c abi got spurious internal dependency).

Do you agree w/ our analysis ?

**Comment by pthierry-ledger** [comment]
replaced by #13486. Can be closed now.

**Pull request closed without merging**